File: mobile/lib/domain/usecases/scan_network_usecase.dart
import '../entities/device.dart';
import '../repositories/device_repository.dart';
import '../repositories/network_repository.dart';

class ScanNetworkUseCase {
  final IDeviceRepository _deviceRepository;
  final INetworkRepository _networkRepository;

  ScanNetworkUseCase(this._deviceRepository, this._networkRepository);

  Future<List<Device>> call() async {
    final scannedDevices = await _networkRepository.scanNetwork();

    List<Device> mergedDevices = [];

    for (var device in scannedDevices) {
      final knownDevice = await _deviceRepository.getDeviceByMac(device.macAddress);

      if (knownDevice != null) {
        mergedDevices.add(device.copyWith(
          isKnown: knownDevice.isKnown,
          category: knownDevice.category,
        ));
      } else {
        mergedDevices.add(device);
      }
    }

    return mergedDevices;
  }
}

File: backend/src/domain/usecases/scan-network.usecase.ts
import { IDeviceRepository } from '../repositories/device.repository';
import { Device } from '../entities/device.entity';

export interface INetworkScanner {
  scanNetwork(): Promise<Device[]>;
}

export class ScanNetworkUseCase {
  constructor(
    private readonly deviceRepository: IDeviceRepository,
    private readonly networkScanner: INetworkScanner,
  ) {}

  async execute(): Promise<Device[]> {
    const scannedDevices = await this.networkScanner.scanNetwork();
    const mergedDevices: Device[] = [];

    for (const device of scannedDevices) {
      const knownDevice = await this.deviceRepository.findByMacAddress(device.macAddress);

      if (knownDevice != null) {
        mergedDevices.push(
          device.copy({
            isKnown: knownDevice.isKnown,
            category: knownDevice.category,
          }),
        );
      } else {
        mergedDevices.push(device);
      }
    }

    return mergedDevices;
  }
}
